README.md
Using solitude for anomaly detection
====================================

-   Srikanth Komala Sheshachala

------------------------------------------------------------------------

Introduction
------------

The R package implements **Isolation forest**, an anomaly detection
method introduced by the paper Isolation based Anomaly Detection ([Liu,
Ting and Zhou](https://dl.acm.org/doi/10.1145/2133360.2133363)).
Isolation forest is grown using
[ranger](https://cran.r-project.org/package=ranger) package.

‘Isolation forest’ is a multivariate outlier detection technique for IID
data containing a mix of numeric and categorical variables.

Usage
-----

    library("solitude")
    packageVersion("solitude")

    ## [1] '1.1.1'

    data("humus", package = "mvoutlier")
    columns_required = setdiff(colnames(humus)
                               , c("Cond", "ID", "XCOO", "YCOO", "LOI")
                               )
    humus2 = humus[ , columns_required]
    str(humus2)

    ## 'data.frame':    617 obs. of  39 variables:
    ##  $ Ag: num  0.124 0.13 0.124 0.309 0.04 0.421 0.235 0.043 0.07 0.086 ...
    ##  $ Al: num  1420 7190 1970 6360 630 2060 1200 1510 3380 1320 ...
    ##  $ As: num  0.743 0.973 1.1 1.24 0.533 1.54 0.836 0.657 1.75 1.04 ...
    ##  $ B : num  2.51 3.45 1.62 1.59 4.33 1.93 2.51 4.46 5.08 2.1 ...
    ##  $ Ba: num  74 65.2 121 63.5 36 111 109 24.5 42.2 51.4 ...
    ##  $ Be: num  0.02 0.08 0.04 0.07 0.01 0.07 0.01 0.05 0.06 0.01 ...
    ##  $ Bi: num  0.108 0.084 0.137 0.135 0.076 0.176 0.145 0.075 0.259 0.112 ...
    ##  $ Ca: num  3900 1900 3240 2470 3500 3340 3010 3090 1760 2810 ...
    ##  $ Cd: num  0.298 0.39 0.295 0.222 0.214 0.327 0.294 0.18 0.295 0.284 ...
    ##  $ Co: num  1.28 3.37 0.98 3.91 0.43 4.61 0.75 0.55 1.68 1.12 ...
    ##  $ Cr: num  3.04 2.02 1.33 17 0.79 3.78 2.01 1.63 10.9 1.71 ...
    ##  $ Cu: num  10.9 12.3 5.59 29.8 6.23 47.3 8.66 5.55 10 9.25 ...
    ##  $ Fe: num  2050 3170 1100 6180 790 3370 1150 1800 4820 970 ...
    ##  $ Hg: num  0.155 0.263 0.168 0.175 0.197 0.239 0.233 0.217 0.238 0.229 ...
    ##  $ K : num  1100 900 1000 900 700 1200 1000 600 800 1000 ...
    ##  $ La: num  2.3 6.4 1.7 5.3 1.2 7 1.5 2.9 6.3 1.2 ...
    ##  $ Mg: num  820 670 790 1180 1770 800 660 1750 820 570 ...
    ##  $ Mn: num  202 22.8 70.3 118 24.4 86.1 1210 16.7 39.2 115 ...
    ##  $ Mo: num  0.662 0.286 0.166 0.28 0.248 0.473 0.185 0.372 0.277 0.15 ...
    ##  $ Na: num  40 120 30 150 410 60 10 320 110 50 ...
    ##  $ Ni: num  9.41 14.8 5.05 59.5 1.98 78 5.7 2.19 11.7 13.2 ...
    ##  $ P : num  743 1030 922 717 795 942 1120 790 779 745 ...
    ##  $ Pb: num  15.9 13.9 19 16.3 14.3 21.7 15.2 11.6 284 14.1 ...
    ##  $ Rb: num  8.14 2.82 4.45 8.13 2.67 5.23 4.94 2.33 4.07 5.35 ...
    ##  $ S : num  1300 1950 1750 965 1860 1410 1880 2210 1260 1250 ...
    ##  $ Sb: num  0.12 0.161 0.25 0.017 0.169 0.217 0.189 0.126 0.809 0.202 ...
    ##  $ Sc: num  0.6 1.2 0.4 1.4 0.2 0.7 0.4 0.6 1.1 0.3 ...
    ##  $ Si: num  630 640 580 690 440 530 650 490 630 560 ...
    ##  $ Sr: num  22.2 34.4 43.4 29.2 46.4 67.9 20.7 59.1 28.8 23.2 ...
    ##  $ Th: num  0.413 0.281 0.246 0.816 0.25 0.432 0.371 0.535 1.13 0.109 ...
    ##  $ Tl: num  0.081 0.068 0.077 0.099 0.06 0.084 0.088 0.042 0.084 0.097 ...
    ##  $ U : num  0.24 0.12 0.042 0.163 0.127 0.112 0.051 0.265 0.221 0.031 ...
    ##  $ V : num  6.79 3.89 2.86 16.1 1.63 7.4 3.55 2.57 13.9 2.76 ...
    ##  $ Y : num  0.8 2.4 0.5 1.9 0.8 1.6 0.5 1.8 1.7 0.4 ...
    ##  $ Zn: num  59.1 18.1 67.6 43.3 41.4 46 103 21.5 23.2 38.5 ...
    ##  $ C : num  39.9 47.5 44.2 19.4 45.8 47.8 45.7 47.3 29.5 46.8 ...
    ##  $ H : num  5.5 6.8 6.3 3 6.1 5.8 6.5 6.4 4.4 6.5 ...
    ##  $ N : num  1.2 1.5 1.5 0.7 1.5 1.3 1.7 1.5 0.7 1 ...
    ##  $ pH: num  3.9 4.1 3.8 4 4.1 3.7 4 4.2 4 3.6 ...

    set.seed(1)
    index = sample(ceiling(nrow(humus2) * 0.5))

    # initiate an isolation forest
    iso = isolationForest$new(sample_size = length(index))
    # fit for attrition data
    iso$fit(humus2[index, ])

    ## INFO  [15:10:47.286] Building Isolation Forest ...  
    ## INFO  [15:10:47.394] done 
    ## INFO  [15:10:47.403] Computing depth of terminal nodes ...  
    ## INFO  [15:10:47.816] done 
    ## INFO  [15:10:47.831] Completed growing isolation forest

    # Obtain anomaly scores
    scores_train = iso$predict(humus2[index, ])
    scores_train[order(anomaly_score, decreasing = TRUE)]

    ##       id average_depth anomaly_score
    ##   1: 169          4.91     0.7258347
    ##   2: 309          6.07     0.6729151
    ##   3: 167          6.34     0.6611618
    ##   4: 106          6.67     0.6470751
    ##   5: 103          6.75     0.6437056
    ##  ---                                
    ## 305: 285          9.00     0.5557972
    ## 306: 286          9.00     0.5557972
    ## 307: 296          9.00     0.5557972
    ## 308: 301          9.00     0.5557972
    ## 309: 306          9.00     0.5557972

    # predict scores for unseen data (50% sample)
    scores_unseen = iso$predict(humus2[-index, ])
    scores_unseen[order(anomaly_score, decreasing = TRUE)]

    ##       id average_depth anomaly_score
    ##   1: 178          6.71     0.6453881
    ##   2: 306          6.77     0.6428659
    ##   3: 260          7.19     0.6254844
    ##   4: 140          7.42     0.6161660
    ##   5: 270          7.42     0.6161660
    ##  ---                                
    ## 304: 282          9.00     0.5557972
    ## 305: 285          9.00     0.5557972
    ## 306: 293          9.00     0.5557972
    ## 307: 295          9.00     0.5557972
    ## 308: 302          9.00     0.5557972

Anomaly detection
-----------------

The paper suggests the following: If the score is closer to 1 for a some
observations, they are likely outliers. If the score for all
observations hover around 0.5, there might not be outliers at all.

By observing the quantiles, we might arrive at the a threshold on the
anomaly scores and investigate the outlier suspects.

    # quantiles of anomaly scores
    quantile(scores_unseen$anomaly_score
             , probs = seq(0.5, 1, length.out = 11)
             )

    ##       50%       55%       60%       65%       70%       75%       80%       85% 
    ## 0.5588889 0.5598015 0.5617045 0.5633013 0.5653084 0.5675263 0.5712422 0.5793897 
    ##       90%       95%      100% 
    ## 0.5889971 0.6012725 0.6453881

The understanding of *why is an observation an anomaly* might require a
combination of domain understanding and techniques like lime (Local
Interpretable Model-agnostic Explanations), Rule based systems etc

Installation
------------

    install.packages("solitude")                  # CRAN version
    devtools::install_github("talegari/solitude") # dev version

------------------------------------------------------------------------


NEWS.md
# solitude 1.1.3

- adds support to sparse matrix input in the form of dgCMatrix

# solitude 1.1.1

- fit gives a message when input dataset has duplicate and does not throw an error

# solitude 1.1.0

- fit exposes max_depth parameter with default value of ceiling(log_2(sample_size))
- path length extension is not done anymore (Thanks to karthi1870)
- ranger's forest object is now public
- fit checks for duplicates in the input dataset

# solitude 1.0.1

- finalized on the API
- isolationForest object no longer calculates and stores anomaly scores on training data. Use predict method for this.
- mtry is no longer exposed
- version 1.0.0 was a faulty release, do not use it

# solitude 0.2.1

- added 'mtry' argument for isolationForest method
- minor optimizations in 'terminalNodesDepth' function

# solitude 0.2.0

- Changes in API: `isolationForest` is now a R6 class with new(), fit() and predict() methods.
- Exported `terminalNodesDepth` function to obtain the depth of terminal nodes of all trees in a ranger forest. 

# solitude 0.1.3

- Changed defaults to avoid bagging in extremely randomized trees (see https://stats.stackexchange.com/questions/406666/what-exactly-is-the-extratrees-option-in-ranger)

- Using `future_lapply` in favor of `furrr::furrr_map`


R/predict.R
#' @name predict.solitude
#' @title Predict method for solitude class
#' @description Predict anomaly score and corrected depth of each observation in
#'   the data using the isolation forest
#' @param object Isolation forest model of class 'solitude'
#' @param data Dataframe to predict on
#' @param type Type of prediction. One among: anomaly_score, depth_corrected
#' @param aggregator Function(as a string, default is 'median') to aggregate the
#'   corrected depths per observation over all trees. This is applicable when
#'   type is anomaly_score
#' @param ... Arguments to be passed to future.apply::future_lapply when future
#'   backend is setup
#' @details The following types of prediction are supported: \itemize{
#'
#'   \item anomaly_score: The thumb rule says: If the score is closer to 1 for a
#'   some obervations, they are likely outliers. If the score for all
#'   observations hover around 0.5, there might not be outliers at all.
#'
#'   \item depth_corrected: This estimates the depth of the observation by
#'   adding a factor of average length unsuccessful search in binary search
#'   tree.
#'
#'   }
#'
#'   See <doi:10.1145/2133360.2133363> for more details.
#'
#'   The predict method supports parallelism via \pkg{future}s.
#' @return Two outputs depending on type argument: \itemize{
#'
#'   \item anomaly_score: A vector(length of number of observations in the data)
#'   of scores. See details for the thumb rule about interpreting them.
#'
#'   \item depth_corrected: A matrix with number of rows equal to number of
#'   observations in the data and number of columns equal to the number of trees
#'   in the model. A value is the estimated depth of an observation in some
#'   tree.
#'
#'   }
#' @examples
#' set.seed(100)
#' index      <- sample.int(150, 75)
#' iris_train <- iris[index, ]
#' iris_test  <- iris[-index, ]
#' mo         <- isolation_forest(iris_train)
#' set.seed(100)
#' index      <- sample.int(150, 100)
#' iris_train <- iris[index, ]
#' iris_test  <- iris[-index, ]
#' mo         <- isolation_forest(iris_train[, 1:4], seed = 101)
#' scores     <- predict(mo, iris_test)
#' summary(scores)
#' with(iris_test
#'      , plot(Sepal.Length
#'             , Sepal.Width
#'             , col = Species
#'             , cex = ifelse(scores > 0.58, 2, 1)
#'             , pch = 20
#'             )
#'      )
#' \dontrun{
#' with(iris_train
#'      , plot(Sepal.Length
#'             , Sepal.Width
#'             , col = Species
#'             , cex = ifelse(predict(mo, iris_train) > 0.6, 2, 1)
#'             , pch = 20
#'             )
#'      )
#' }
#' @importFrom stats predict
#' @importFrom data.table :=
#' @export
predict.solitude <- function(object
                             , data
                             , type       = "anomaly_score"
                             , aggregator = "median"
                             , ...
                             ){

  if(!(type %in% c("anomaly_score", "depth_corrected"))){
    stop("type has to be among: anomaly_score, depth_corrected")
  }

  res <- switch(type
                , depth_corrected = depth_corrected(object, data, ...)
                , anomaly_score   = anomaly_score(object, data, aggregator, ...)
                )

  return(res)
}

#' @name depth_corrected
#' @title depth_corrected
#' @description depth_corrected
#' @param object isolation forest model
#' @param newdata dataframe to predict
#' @param ... Arguments to be passed to future.apply::future_lapply
depth_corrected <- function(object, newdata, ...){

  args_future <- list(...)

  num_trees <- object[["forest"]][["num.trees"]]

  tnm <- predict(object[["forest"]]
                 , data = newdata
                 , type = "terminalNodes"
                 )[["predictions"]] + 1L


  get_corrected_depths <- function(x){
    depths <- c(
      depth_terminalNodes(ranger::treeInfo(object[["forest"]], x))
      , average_path_length(table(tnm[, x]))
      )

    tapply(depths, names(depths), sum)
  }

  corrected_depths <-
    Reduce(
      cbind
      , fastDoCall(
          future.apply::future_lapply
          , c(
              list(1:num_trees
                   , function(x) as.numeric(get_corrected_depths(x)[as.character(tnm[, x])])
                   )
             , args_future
             )
          )
    )

  return(corrected_depths)
}

#' @name anomaly_score
#' @title anomaly_score
#' @description anomaly_score
#' @param object isolation forest model
#' @param newdata dataframe to predict
#' @param aggregator aggregator
#' @param ... Arguments to be passed to future.apply::future_lapply in
#'   'depth_corrected'
anomaly_score <- function(object, newdata, aggregator = "median", ...){

  corrected_depths <- depth_corrected(object, newdata, ...)
  res <- compute_anomaly(apply(corrected_depths
                               , 1
                               , eval(as.symbol(aggregator))
                               )
                         , nrow(newdata)
                         )
  return(res)
}

#' @name depth_terminalNodes
#' @title depth_terminalNodes
#' @description depth_terminalNodes
#' @param treelike A single ranger tree extracted from 'ranger::treeInfo'
depth_terminalNodes <- function(treelike){

  nodeID     <- NULL
  leftChild  <- NULL
  rightChild <- NULL

  data.table::setDT(treelike)
  dropThese <- setdiff(colnames(treelike)
                       , c("nodeID", "leftChild", "rightChild")
                       )
  treelike[, c(dropThese) := NULL]
  melted    <- data.table::melt(treelike
                                , id.vars      = "nodeID"
                                , measure.vars = c("leftChild", "rightChild")
                                )
  value       <- NULL
  edgeMat     <- as.matrix(melted[!is.na(value), c("nodeID", "value")]) + 1L
  treegraph   <- igraph::graph_from_edgelist(edgeMat)
  tnValues    <- treelike[is.na(leftChild) & is.na(rightChild), nodeID] + 1L
  depths      <- igraph::distances(treegraph
                                   , v    = 1
                                   , to   = tnValues
                                   , mode = "out"
                                   )
  dim(depths) <- NULL
  names(depths) <- tnValues
  return(depths)
}

#' @name average_path_length
#' @title average_path_length
#' @description average_path_length
#' @param n n
average_path_length <- Vectorize(
  function(n){
    ifelse(n == 1, 0, 2 * ( harmonic(n - 1) - ((n - 1)/n) ))
  }
  , vectorize.args = "n"
)

#' @name harmonic_approx
#' @title harmonic_approx
#' @description harmonic_approx
#' @param n n
harmonic_approx <- function(n){ log(n) + 0.577216 }

#' @name harmonic_exact
#' @title harmonic_exact
#' @description harmonic_exact
#' @param n n
harmonic_exact <-
  Vectorize(
    function(n){
      sum(1/seq(1,n))
    }
    , vectorize.args = "n"
    )

#' @name harmonic
#' @title harmonic
#' @description harmonic
#' @param n n
harmonic <- Vectorize(
  function(n){
    if(n < 11){
      harmonic_exact(n)
    } else {
      harmonic_approx(n)
    }
  }
  , vectorize.args = "n"
  )

#' @name compute_anomaly
#' @title compute_anomaly
#' @description compute_anomaly
#' @param pathLength pathLength
#' @param dataSize dataSize
compute_anomaly <- Vectorize(
  function(pathLength, dataSize){
    2^( -( pathLength / average_path_length(dataSize) ) )
  }
  , vectorize.args = "pathLength"
  )


R/fastDoCall.R
#' An alternative to the internal \code{do.call}
#'
#' The \code{\link[base]{do.call}} can be somewhat slow, especially when working
#' with large objects. This function is based upon the suggestions from Hadley
#' Wickham on the R mailing list, see
#' \href{http://r.789695.n4.nabble.com/call-do-call-expression-too-big-td3574335.html}{here}.
#' Also thanks to \emph{Tommy} at StackOverflow for
#' \href{https://stackoverflow.com/questions/10022436/do-call-in-combination-with}{suggesting}
#' how to handle double and triple colon operators, \code{::}, further enhancing
#' the function.
#'
#' @inheritParams base::do.call
#'
#' @section Note:
#'
#'   Note that this is an verbaim copy of fastDoCall function from Gmisc
#'   package. All credits for Max Gordon.

fastDoCall <- function(what, args, quote = FALSE, envir = parent.frame()){
  if (quote)
    args <- lapply(args, enquote)

  if (is.null(names(args)) ||
        is.data.frame(args)){
    argn <- args
    args <- list()
  }else{
    # Add all the named arguments
    argn <- lapply(names(args)[names(args) != ""], as.name)
    names(argn) <- names(args)[names(args) != ""]
    # Add the unnamed arguments
    argn <- c(argn, args[names(args) == ""])
    args <- args[names(args) != ""]
  }

  if (class(what) == "character"){
    if(is.character(what)){
      fn <- strsplit(what, "[:]{2,3}")[[1]]
      what <- if(length(fn)==1) {
        get(fn[[1]], envir=envir, mode="function")
      } else {
        get(fn[[2]], envir=asNamespace(fn[[1]]), mode="function")
      }
    }
    call <- as.call(c(list(what), argn))
  }else if (class(what) == "function"){
    f_name <- deparse(substitute(what))
    call <- as.call(c(list(as.name(f_name)), argn))
    args[[f_name]] <- what
  }else if (class(what) == "name"){
    call <- as.call(c(list(what, argn)))
  }

  eval(call,
       envir = args,
       enclos = envir)
}


R/terminal_nodes.R
#' @name terminalNodesDepth
#' @title Depth of each terminal node of all trees in a ranger model
#' @description Depth of each terminal node of all trees in a ranger model is
#'   returned as a three column tibble with column names: 'id_tree',
#'   'id_node', 'depth'. Note that root node has the node_id = 0.
#' @param model A ranger model
#' @details This function may be parallelized using a future backend.
#' @return A tibble with three columns: 'id_tree', 'id_node', 'depth'.
#' @examples
#' rf = ranger::ranger(Species ~ ., data = iris, num.trees = 100)
#' terminalNodesDepth(rf)
#' @export
terminalNodesDepth = function(model){

  id_node   = NULL
  depth     = NULL
  num_trees = model[["num.trees"]]

  # iterate over trees
  tnd = future.apply::future_lapply(
    1:num_trees
    , function(x) terminalNodesDepthPerTree(ranger::treeInfo(model, x))
    )
  tnd = data.table::rbindlist(tnd, idcol = "id_tree")

  tnd[ , id_node := as.integer(id_node)]
  tnd[ , depth   := as.integer(depth)]

  return(tnd[])
}


#' @name terminalNodesDepthPerTree
#' @title Depth of each terminal node of a single tree in a ranger model
#' @description Depth of each terminal node of a single tree in a ranger model.
#'   Note that root node has the id_node = 0.
#' @param treelike Output of `ranger::treeInfo`
#' @return data.table with two columns: id_node and depth
#' @examples
#' \dontrun{
#'   rf = ranger::ranger(Species ~ ., data = iris)
#'   terminalNodesDepthPerTree(ranger::treeInfo(rf, 1))
#' }
terminalNodesDepthPerTree = function(treelike){

  treelike  = data.table::as.data.table(treelike)
  melted    = data.table::melt(
    treelike[, c("nodeID", "leftChild", "rightChild")]
    , id.vars      = "nodeID"
    , measure.vars = c("leftChild", "rightChild")
    , na.rm        = TRUE
    )

  # create graph in 1-index mode
  edgeMat     = as.matrix(melted[ , c("nodeID", "value")]) + 1L
  treegraph   = igraph::graph_from_edgelist(edgeMat, directed = TRUE)

  # get depths of terminal nodes
  tnValues    = treelike[["nodeID"]][treelike[["terminal"]]]
  depths      = igraph::distances(treegraph
                                   , v    = 1L
                                   , to   = tnValues + 1L # to 1-index mode
                                   , mode = "out"
                                   )
  dim(depths) = NULL

  res = data.table::data.table(id_node = tnValues, depth = depths)
  return(res)
}


R/isolation_forest.R
#' @name isolationForest
#' @aliases isolation_forest
#' @title Grow an isolation forest
#' @description Implements isolation forest  from Isolation-Based Anomaly
#'   Detection (<doi:10.1145/2133360.2133363>)
#' @details This function internally uses \code{\link[ranger]{ranger}}.
#'   Arguments may be passed to it except these: dependent.variable.name, data,
#'   mtry, min.node.size, splitrule, num.random.splits,
#'   respect.unordered.factors. Pass 'seed' argument for reproducibility
#' @param dataset Dataframe with column names
#' @param ... Optional arguments passed to \code{\link[ranger]{ranger}}
#' @return An object(list) of class 'solitude'. The isolationforest is in the
#'   'forest' slot
#' @examples
#' mo <- isolationForest(iris, seed = 1)
#' @export
isolationForest <- function(dataset, ...){

  columnNames  <- colnames(dataset)
  responseName <- columnNames[[1]]
  while(deparse(substitute(responseName)) %in% columnNames){
    responseName <- sample(c(letters, LETTERS), 20, replace = TRUE)
  }

  arguments <- list(...)
  arg_names <- names(arguments)

  fixed_names <- c("dependent.variable.name"
                   , "data"
                   , "mtry"
                   , "min.node.size"
                   , "splitrule"
                   , "num.random.splits"
                   )
  if(any(fixed_names %in% arg_names)){
    stop(
      paste0("These optional arguments of ranger::ranger should not be specified: "
             , toString(fixed_names)
             )
      )
  }

  if("seed" %in% names(arguments)){
    seed <- arguments[["seed"]]
  } else {
    seed <- sample.int(1e5, 1)
  }

  if(!("replace" %in% arg_names)){
    arguments[["replace"]] <- FALSE
  }

  if(!("sample.fraction" %in% arg_names)){
    arguments[["sample.fraction"]] <- 1
  }

  if(!("respect.unordered.factors" %in% arg_names)){
    arguments[["respect.unordered.factors"]] <- "partition"
  }

  set.seed(seed)
  dataset[[deparse(substitute(responseName))]] <- sample.int(nrow(dataset))

  # build a isolation forest
  iso <- fastDoCall(ranger::ranger
    , c(
        list(
          dependent.variable.name     = deparse(substitute(responseName))
          , data                      = dataset
          , mtry                      = 1
          , min.node.size             = 1
          , splitrule                 = "extratrees"
          , num.random.splits         = 1
          )
        , arguments
        )
    )

  return(structure(list(forest = iso), class = "solitude"))
}

#' @export
isolation_forest <- isolationForest

R/isolationForest_r6.R
#' @name isolationForest
#' @title Fit an Isolation Forest
#' @description 'solitude' class implements the isolation forest method
#'   introduced by paper Isolation based Anomaly Detection (Liu, Ting and Zhou
#'   <doi:10.1145/2133360.2133363>). The extremely randomized trees (extratrees)
#'   required to build the isolation forest is grown using
#'   \code{\link[ranger]{ranger}} function from \pkg{ranger} package.
#' @section Design: \code{$new()} initiates a new 'solitude' object. The
#'   possible arguments are:
#'
#'   \itemize{
#'
#'   \item \code{sample_size}: (positive integer, default = 256) Number of
#'   observations in the dataset to used to build a tree in  the forest
#'
#'   \item \code{num_trees}: (positive integer, default = 100) Number of trees
#'   to be built in the forest
#'
#'   \item \code{replace}: (boolean, default = FALSE) Whether the sample of
#'   observations should be chosen with replacement when sample_size is less
#'   than the number of observations in the dataset
#'
#'   \item \code{seed}: (positive integer, default = 101) Random seed for the
#'   forest
#'
#'   \item \code{nproc}: (NULL or a positive integer, default: NULL, means use
#'   all resources) Number of parallel threads to be used by ranger
#'
#'   \item \code{respect_unordered_factors}: (string, default: "partition")See
#'   respect.unordered.factors argument in \code{\link[ranger]{ranger}}
#'
#'   \item \code{max_depth}: (positive number, default:
#'   ceiling(log2(sample_size))) See max.depth argument in
#'   \code{\link[ranger]{ranger}}
#'
#'   }
#'
#'   \code{$fit()} fits a isolation forest for the given dataframe or sparse matrix, computes
#'   depths of terminal nodes of each tree and stores the anomaly scores and
#'   average depth values in \code{$scores} object as a data.table
#'
#'   \code{$predict()} returns anomaly scores for a new data as a data.table
#'
#' @section Details:
#'
#'   \itemize{
#'
#'   \item Parallelization: \code{\link[ranger]{ranger}} is parallelized and by
#'   default uses all the resources. This is supported when nproc is set to
#'   NULL. The process of obtaining depths of terminal nodes (which is excuted
#'   with \code{$fit()} is called) may be parallelized separately by setting up
#'   a \pkg{future} backend.
#'
#'   }
#'
#' @examples
#' \dontrun{
#' library("solitude")
#' library("tidyverse")
#' library("mlbench")
#'
#' data(PimaIndiansDiabetes)
#' PimaIndiansDiabetes = as_tibble(PimaIndiansDiabetes)
#' PimaIndiansDiabetes
#'
#' splitter   = PimaIndiansDiabetes %>%
#'   select(-diabetes) %>%
#'   rsample::initial_split(prop = 0.5)
#' pima_train = rsample::training(splitter)
#' pima_test  = rsample::testing(splitter)
#'
#' iso = isolationForest$new()
#' iso$fit(pima_train)
#'
#' scores_train = pima_train %>%
#'   iso$predict() %>%
#'   arrange(desc(anomaly_score))
#'
#' scores_train
#'
#' umap_train = pima_train %>%
#'   scale() %>%
#'   uwot::umap() %>%
#'   setNames(c("V1", "V2")) %>%
#'   as_tibble() %>%
#'   rowid_to_column() %>%
#'   left_join(scores_train, by = c("rowid" = "id"))
#'
#' umap_train
#'
#' umap_train %>%
#'   ggplot(aes(V1, V2)) +
#'   geom_point(aes(size = anomaly_score))
#'
#' scores_test = pima_test %>%
#'   iso$predict() %>%
#'   arrange(desc(anomaly_score))
#'
#' scores_test
#' }
#' @export

isolationForest = R6::R6Class(
  "solitude"
  ,
  # Implementation:
  # 'solitude' class implements the isolation forest method defined by
  # Liu, Fei Tony, Ting, Kai Ming and Zhou, Zhi-Hua.
  # "Isolation-based anomaly detection." ACM Transactions on Knowledge Discovery   # from Data (T KDD). <doi:10.1145/2133360.2133363>
  #
  # Plan:
  # 1. Build 'extratrees' forest using ranger.
  # 2. Obtain terminal node depths by parsing each tree.
  # 3. Compute anomaly scores.
  #
  # Design:
  # 1. 'initialize' method sets the params required for forest growth.
  # 2. Calling 'fit' on some dataset does these things:
  #    - Grow 'extratrees' forest.
  #    - Compute and store the depth of each terminal node in every tree.
  #    - Compute depth or pathlength for each observation.
  #    - Compute anomaly score for each observation and store it in 'scores'.
  # 3. 'predict' method computes anomaly scores on new data.
  #
  public = list(

    sample_size                 = NULL
    , num_trees                 = NULL
    , replace                   = NULL
    , seed                      = NULL
    , nproc                     = NULL
    , respect_unordered_factors = NULL
    , max_depth                 = NULL
    , forest                    = NULL
    , scores = NULL
    , status = "not_initialized"
    ,
    # intialize arguments required for fitting extratrees via ranger
    initialize = function(sample_size                 = 256
                          , num_trees                 = 100
                          , replace                   = FALSE
                          , seed                      = 101
                          , nproc                     = NULL
                          , respect_unordered_factors = NULL
                          , max_depth                 = ceiling(log2(sample_size))
                          ){

      stopifnot(is_integerish(sample_size) && length(sample_size) == 1)
      stopifnot(0 < sample_size)
      stopifnot(is_integerish(num_trees) && length(num_trees) == 1)
      stopifnot(0 < num_trees)
      stopifnot(is.logical(replace) && length(replace) == 1)
      stopifnot(is_integerish(seed) && length(seed) == 1)
      stopifnot(is.null(nproc) ||
                  (is_integerish(nproc) && length(nproc == 1) && nproc >= 1)
                )
      stopifnot(is.null(respect_unordered_factors) ||
                  (is.character(respect_unordered_factors) &&
                  length(respect_unordered_factors) == 1)
                )
      stopifnot(is_integerish(max_depth) &&
                length(max_depth) == 1 &&
                max_depth > 0
                )

      self$sample_size               = sample_size
      self$num_trees                 = num_trees
      self$replace                   = replace
      self$seed                      = seed
      self$nproc                     = nproc
      self$respect_unordered_factors = respect_unordered_factors
      self$max_depth                 = max_depth
      self$status                    = "not_trained"
    }
    ,
    fit = function(dataset){

      is_sparse = grepl("dg.Matrix", class(dataset)[[1]])
      # check if any rows are duplicated
      if (is_sparse) {
        lgr::lgr$info("sparse dataset detected, skipping duplication check")
      } else if (anyDuplicated(dataset) > 0) {
        lgr::lgr$info("dataset has duplicated rows")
      }

      # create new fit
      if(self$status == "trained"){
        self$status = "not_trained"
        lgr::lgr$info("Retraining ... ")
      }

      # create a new 'y' column with jumbled 1:n
      columnNames  = colnames(dataset)
      nr           = nrow(dataset)

      set.seed(self$seed)
      yy = sample.int(nrow(dataset))

      # deduce sample_fraction
      stopifnot(self$sample_size <= nr)
      private$sample_fraction = self$sample_size/nr

      # build a extratrees forest
      lgr::lgr$info("Building Isolation Forest ... ")
      self$forest = ranger::ranger(
        x = dataset
        , y = yy
        , mtry                      = ncol(dataset) - 1L
        , min.node.size             = 1L
        , splitrule                 = "extratrees"
        , num.random.splits         = 1L
        , num.trees                 = self$num_trees
        , replace                   = self$replace
        , sample.fraction           = private$sample_fraction
        , respect.unordered.factors = self$respect_unordered_factors
        , num.threads               = self$nproc
        , seed                      = self$seed
        , max.depth                 = self$max_depth
        )
      lgr::lgr$info("done")

      # compute terminal nodes depth
      lgr::lgr$info("Computing depth of terminal nodes ... ")
      private$terminal_nodes_depth = terminalNodesDepth(self$forest)
      lgr::lgr$info("done")

      # set phi -- sample size used for tree building
      private$phi = floor(private$sample_fraction * nr)

      # create path length extend dataframe
      tnm = stats::predict(self$forest
                           , dataset
                           , type        = "terminalNodes"
                           , num.threads = self$nproc
                           )[["predictions"]]

      tnm = data.table::as.data.table(tnm)
      data.table::setnames(tnm, colnames(tnm), as.character(1:ncol(tnm)))
      tnm[, id := .I]
      tnm = data.table::melt(tnm
                             , id.vars         = "id"
                             , variable.name   = "id_tree"
                             , value           = "id_node"
                             , variable.factor = FALSE
                             )
      id_tree = NULL
      id_node = NULL

      tnm[, id_tree := as.integer(id_tree)]
      tnm[, id_node := as.integer(id_node)]

      # update train status
      self$status = "trained"
      lgr::lgr$info("Completed growing isolation forest")
    }
    ,
    predict = function(data){

      tnm = stats::predict(self$forest
                           , data
                           , type        = "terminalNodes"
                           , num.threads = self$nproc
                           )[["predictions"]]

      tnm = data.table::as.data.table(tnm)
      data.table::setnames(tnm, colnames(tnm), as.character(1:ncol(tnm)))
      tnm[, id := .I]
      tnm = data.table::melt(tnm
                             , id.vars         = "id"
                             , variable.name   = "id_tree"
                             , value           = "id_node"
                             , variable.factor = FALSE
                             )
      id_tree = NULL
      id_node = NULL

      tnm[, id_tree := as.integer(id_tree)]
      tnm[, id_node := as.integer(id_node)]

      obs_depth = merge(tnm
                        , private$terminal_nodes_depth
                        , by = c("id_tree", "id_node")
                        )

      average_depth = NULL
      depth         = NULL
      id            = NULL
      anomaly_score = NULL

      scores = obs_depth[ , .(average_depth = mean(depth)), by = id][
        order(id)]
      scores[, anomaly_score :=
               private$computeAnomaly(average_depth, private$phi)][]

      return(scores)
    }
  )
  ,
  private = list(

    terminal_nodes_depth       = NULL
    , phi                      = NULL
    , sample_fraction          = NULL
    ,
    pathLengthNormalizer = function(phi){

      res = 0

      if(phi == 2){
        res = 1
      }
      if(phi > 2){
        res = (2 * private$harmonic(phi - 1)) - (2 * (phi - 1)/phi)
      }

      return(res)

    }
    ,
    computeAnomaly = function(depth, data_size){

      den = private$pathLengthNormalizer(data_size)
      return(2^( -( depth / den ) ))
    }
    ,
    harmonic = function(n){
      ifelse(n > 11, log(n) + 0.577216, sum(1/seq(1,n)))
    }
  )
)
